### ISSUE TITLE
Error should be raised when pessimistically locking with reader role

### ISSUE THREAD

**Issue by joshuay03** [rails/rails#53758]
### Steps to reproduce

In our development environment, we use the same PostgreSQL database for both the writer and reader roles. While testing some automatic role-switching middleware and relying on our test suite to detect writes in read-only mode, we encountered a hanging test. We traced the issue to a pessimistic lock being acquired using with_lock (`SELECT ... FOR UPDATE`) while using the reader role, followed by switching to the writer role to update the same row. This resulted in a deadlock, as the `UPDATE` transaction waited on the `SELECT ... FOR UPDATE` lock, and vice versa.

From my understanding, PostgreSQL would have raised an error (`ERROR:  cannot execute SELECT FOR UPDATE in a read-only transaction`) if this had been attempted in a read-only role, as it would in our production environment. Therefore, I believe Rails could potentially raise an error in this scenario as well (maybe with differing behavior based on the adapter).

Please let me know if a PR to raise an error in this case would be accepted and I can send one 👍🏽

```ruby
# frozen_string_literal: true

require "bundler/inline"

gemfile(true) do
  source "https://rubygems.org"

  gem "rails", github: "rails/rails", branch: "main"

  gem "pg", "~> 1.1"
end

require "active_record/railtie"
require "minitest/autorun"

ActiveRecord::Base.logger = Logger.new(STDOUT)

# Note: This script requires a local PostgresQL database `primary`
config = YAML.load(<<~YAML)
  development:
    primary:
      adapter: postgresql
      database: primary
    replica:
      adapter: postgresql
      database: primary
      replica: true
YAML
ActiveRecord::Base.configurations = config

ActiveRecord::Base.establish_connection :primary

ActiveRecord::Schema.define do
  create_table :posts, force: :cascade do |t|
    t.text :title
  end
end

class ApplicationRecord < ActiveRecord::Base
  primary_abstract_class

  connects_to database: { writing: :primary, reading: :replica }
end

class Post < ApplicationRecord
end

class BugTest < ActiveSupport::TestCase
  def test_pessimistic_lock_with_database_switch
    post = Post.create!(title: "original")
    ApplicationRecord.connected_to(role: :reading) do
      post.with_lock do
        ApplicationRecord.connected_to(role: :writing) do
          post.update!(title: "updated")
        end
      end
    end
  end
end
```

### Expected behavior

An error should be raised telling me that a lock cannot be obtained when in readonly mode, similar to: https://github.com/rails/rails/blob/07bf01aba4904ff2f9eb5cfd11b33c5d46db26a8/activerecord/lib/active_record/connection_adapters/abstract_adapter.rb#L208

### Actual behavior

It hangs indefinitely.

### System configuration

**Rails version**: Edge

**Ruby version**: 3.4.0preview2

**Comment by fatkodima** [comment]
Agreed, makes sense to me as locking rows is usually precedes some follow up updates anyway (which need the up to date data, which is not the case for replicas).

**Comment by joshuay03** [comment]
> Agreed, makes sense to me as locking rows is usually precedes some follow up updates anyway (which need the up to date data, which is not the case for replicas).

I suspect it could use more test coverage, but here’s what I’ve put together: https://github.com/rails/rails/pull/53974

**Comment by joshuay03** [comment]
> I suspect it could use more test coverage, but here’s what I’ve put together: [#53974](https://github.com/rails/rails/pull/53974)

Went with an alternative implementation: https://github.com/rails/rails/pull/54580
